### PR TITLE
Fix OSS-Fuzz issues

### PR DESCRIPTION
**Pull request by xhanulik**
This PR contains fixes for recent issues reported by OSS-Fuzz

##### Checklist
- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

File: src/pkcs15init/profile.c [hunk 1]
```diff
@@ -1985,24 +1985,40 @@ static struct block	root_ops = {
    "root",		process_block,	NULL,		root_blocks
 };
 
+static void
+get_inner_word(char *str, char word[64]) {
```
Comment (xhanulik) [inline src/pkcs15init/profile.c:1989]: I'm not sure whether the macros can be nested inside another expression; the problematic one found by OSS-Fuzz looks like `odf-size=($odf-size�v$` (with some non-alphanumeric byte).

File: src/pkcs15init/pkcs15-lib.c [hunk 2]
```diff
@@ -1674,6 +1675,8 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 		/* can be allocated in driver-specific generate_key() */
 		free(pubkey_args.key.u.ec.ecpointQ.value);
 		pubkey_args.key.u.ec.ecpointQ.value = NULL;
+		if (r < 0 && alg_id_free)
+			free(pubkey->alg_id);
```
Comment (Jakuje) [inline src/pkcs15init/pkcs15-lib.c:1679]: 1) I think this should not be in the condition `(algorithm == SC_ALGORITHM_EC)` as this can happen also for other algorithms.
2) I think in regards to the recent changes to this code I did with valgrind, I think the `alg_id` should be copied in all cases and free here in all cases to simplify the memory management, similarly as done with the other parts of pubkey structure.

File: src/pkcs15init/profile.c (hunk 1 above)
Comment (Jakuje) [reply src/pkcs15init/profile.c:1989]: I think this is mostly valid use case for example in the myeid profile object, there are macros used as part of the strings (indeed unless it is recursive).

File: src/pkcs15init/profile.c [hunk 3]
```diff
@@ -1985,24 +1985,40 @@ static struct block	root_ops = {
...
+	char *inner = NULL;
+	size_t len = 0;
+
+	inner = str;
+	while (isalnum((unsigned char)*inner) || *inner == '-' || *inner == '_') {
```
Comment (Jakuje) [inline src/pkcs15init/profile.c:1994]: This condition is already used in `__expr_get()` to filter the macro names. I would prefer to move the condition to separate function to have these two uses in sync and make sure the do not diverge.

File: src/pkcs15init/profile.c [hunk 4]
```diff
@@ -1985,24 +1985,40 @@ static struct block	root_ops = {
...
+		inner++;
+		len++;
+	}
+	len = len >= 64 ? 63 : len;
```
Comment (Jakuje) [inline src/pkcs15init/profile.c:1998]: using `sizeof(word)` instead of magic number would it make a bit more readable.

File: src/pkcs15init/profile.c [hunk 5]
```diff
@@ -1985,24 +1985,40 @@ static struct block	root_ops = {
...
+	memcpy(word, str, len);
+	word[len] = '\0';
+}
+
 static int
-check_macro_reference_loop(char *start, char *str, struct state *cur, int depth) {
+check_macro_reference_loop(const char *start_name, char *value, sc_profile_t *profile, int depth) {
```
Comment (Jakuje) [inline src/pkcs15init/profile.c:2004]: I would propose a short doc string here with at least information what return values we can expect in which cases.
Ideally introducing also some unit tests to check the corner cases if possible as it is getting more complicated.

File: src/pkcs15init/profile.c (hunk 4 above)
Comment (xhanulik) [reply src/pkcs15init/profile.c:1998]: This would return only size of `char *`, but will use macro instead, thanks.

File: src/pkcs15init/profile.c (hunk 1 above)
Comment (xhanulik) [reply src/pkcs15init/profile.c:1989]: Ok, thanks for checking.

File: src/pkcs15init/pkcs15-lib.c [hunk 6]
```diff
@@ -1664,6 +1664,7 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 
 err:
 	sc_pkcs15_free_object(object);
+	free(pubkey->alg_id);
```
Comment (Jakuje) [inline src/pkcs15init/pkcs15-lib.c:1667]: This is not what I meant -- now you are freeing it even if it was not copied, which might cause double frees.

I had in my mind modifying the block from line 1629 to happen in any case to make sure we do not modify the existing sc_algorith_id structure pointed by `pubkey_args` if it is provided by caller and not share it with the newly created `pubkey` object to avoid the double-free.

Comment (Jakuje) [reply src/pkcs15init/pkcs15-lib.c]: My bad. You are right!

File: src/pkcs15init/pkcs15-lib.c [hunk 7]
```diff
@@ -1664,6 +1665,8 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 
 err:
 	sc_pkcs15_free_object(object);
+	if (r < 0 && alg_id_free)
+		free(pubkey->alg_id);
```
Comment (Jakuje) [inline src/pkcs15init/pkcs15-lib.c:1669]: I think the `pubkey->alg_id` should be freed regardless the return of this function as it is pointer to the `pubkey_args`, which is local to this function. It is passed to the `sc_pkcs15init_store_public_key()`, but if it needs it for something, it should copy it out similarly as with the ec params discussed in #2790.

And given that the `pubkey_args` is correctly zeroized in the beginning of this function, I think the `alg_id_free` is not needed either. It will be allocated either here or in the generate_key provided by the card driver and there is no card-specific free after this ...

File: src/pkcs15init/pkcs15-lib.c [hunk 8]
```diff
@@ -1664,6 +1664,7 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 
 err:
 	sc_pkcs15_free_object(object);
+	free(pubkey_args.key.alg_id);
```
[the remark this answers is not in the document]
Comment (frankmorgner) [reply src/pkcs15init/pkcs15-lib.c:1667]: Sorry, I just noticed your above discussion on this. I'll just wait a little while to see if you get to a conclusion...

[the remark this answers is not in the document]
Comment (Jakuje) [reply src/pkcs15init/pkcs15-lib.c:1667]: I think in this case it is a not the caller, but the `generate_key()` callback of the pkcs15 emulator.

I did not go checking deeper if and when this is initialized in the drivers, but it looks like there is some different cleaning of this -- it should be easy to get traces with valgrind now (On the side note, it is now not executed in CI now because the regular run already crashes -- we might want to either execute the valgrind in the separate job or even if it would fail -- not sure what would be the best).

I think the it should be investigated slightly deeper rather than implementing fast solution using local variables like proposed initially. On the other hand if it is double free, setting just NULL after the free might solve the issue.

[the remark this answers is not in the document]
Comment (xhanulik) [reply src/pkcs15init/pkcs15-lib.c:1667]: The double-free happens because `sc_pkcs15init_store_public_key()` in https://github.com/OpenSC/OpenSC/blob/b031789dccbdc46d151354719f0d527fe100cf76/src/pkcs15init/pkcs15-lib.c#L1654-L1667 releases the `alg_id` via `sc_pkcs15_erase_pubkey()`. Setting `NULL` there should work fine.

[the remark this answers is not in the document]
Comment (xhanulik) [reply src/pkcs15init/pkcs15-lib.c:1667]: If I understand correctly, `sc_pkcs15_erase_pubkey()` in `sc_pkcs15init_store_public_key()` is called on _(shallow) copy_,
https://github.com/OpenSC/OpenSC/blob/b031789dccbdc46d151354719f0d527fe100cf76/src/pkcs15init/pkcs15-lib.c#L1858-L1860
so the `0/NULL`s are not propagated outside. The second call to  `sc_pkcs15_erase_pubkey()` then causes another double-free.

**Review by Jakuje — APPROVED** [review]

**Comment by frankmorgner** [comment]
Thank you!
